### PR TITLE
Rework enums to use the same name mapping as message fields.

### LINKED ISSUE #247: Use `FieldNameMap` on `Enum` for textual names
As discussed in PR #243:

https://github.com/apple/swift-protobuf/pull/243/files/e397b41fa359a4aa7f6fec465b4a2d357b661d65#diff-7ed1330a33d576e5f9e1354cfb77e795L23

we should be able to reuse `FieldNameMap` with `Enum` to provide number ⇔ name mapping.  A few notes:
* Maybe rename `FieldNameMap` to just `NameMap`?
* Can we use the exact same `ProtoNameProviding` protocol here?

Among other things, this would work well with a future generator option to omit the textual names.

### PR DESCRIPTION
**Pull request by allevato**
This cuts a decent amount of codegen and makes the access to enum names internal only. Users shouldn't need access to those any more than they need the field names. (They *can* still get at them because the name map is public, of course, but it's not convenient and not meant to be.)

Fixes #247.

### CODE REVIEW

File: Sources/SwiftProtobuf/Enum.swift [hunk 1]
```diff
@@ -15,21 +15,48 @@
...
+  public var hashValue: Int {
+    return rawValue
+  }
+
+  /// Internal convenience property representing the name of the enum value (or
+  /// `nil` if it is an `UNRECOGNIZED` value or doesn't provide names).
+  ///
+  /// Since the text format and JSON names are always identical, we don't need
+  /// to distinguish them.
+  internal var name: StaticString? {
```
Comment (thomasvl) [inline Sources/SwiftProtobuf/Enum.swift]: So what does this mean as far as an enum having a case of "name"?  i.e. - should we prefix these?

Comment (allevato) [inline Sources/SwiftProtobuf/Enum.swift]: Cases won't collide with instance properties. (For all intents and purposes, a case is a static method that returns an instance of the enum.)

File: Sources/SwiftProtobuf/NameMap.swift [hunk 2]
```diff
@@ -0,0 +1,155 @@
+// Sources/SwiftProtobuf/fieldNameMap.swift - Bidirectional field number/name mapping
```
Comment (thomasvl) [inline Sources/SwiftProtobuf/NameMap.swift]: Update comment for new name.

Looks like git didn't think this came from the other file.  :(

Comment (allevato) [inline Sources/SwiftProtobuf/NameMap.swift]: Oops, find-and-replace gone awry too.

There are significant enough changes within the file to defeat git's crappy rename detection :(

File: Sources/SwiftProtobuf/type.pb.swift [hunk 3]
```diff
@@ -71,39 +76,11 @@ public enum Google_Protobuf_Syntax: SwiftProtobuf.Enum {
...
-      }
-    }
-  }
-
-  public var _protobuf_jsonName: String? {
-    get {
-      switch self {
-      case .proto2: return "SYNTAX_PROTO2"
-      case .proto3: return "SYNTAX_PROTO3"
-      case .UNRECOGNIZED: return nil
```
Comment (thomasvl) [inline Sources/SwiftProtobuf/type.pb.swift]: Did we have a test case that covered this one?

**Comment by tbkka** [comment]
Looks like this also fixes #177?

File: Sources/protoc-gen-swift/EnumCaseGenerator.swift [hunk 4]
```diff
@@ -0,0 +1,56 @@
+// Sources/protoc-gen-swift/EnumGenerator.swift - Enum case logic
```
Comment (thomasvl) [inline Sources/protoc-gen-swift/EnumCaseGenerator.swift]: Update

Comment (allevato) [inline Sources/protoc-gen-swift/EnumCaseGenerator.swift]: And this is why it's not a great idea to include the name/path of a source file in the file...

**Comment by allevato** [comment]
> Looks like this also fixes #177?

I believe it does (in part; @thomasvl already did some of the work, like removing the quotes from the JSON names, IIRC).

File: Sources/SwiftProtobuf/type.pb.swift (hunk 3 above)
Comment (allevato) [inline Sources/SwiftProtobuf/type.pb.swift]: Looks like there are some in `Test_Enum`.

File: Reference/conformance/conformance.pb.swift [hunk 5]
```diff
@@ -88,43 +94,12 @@ enum Conformance_WireFormat: SwiftProtobuf.Enum {
...
-  init?(jsonName: String) {
-    switch jsonName {
-    case "UNSPECIFIED": self = .unspecified
-    case "PROTOBUF": self = .protobuf
-    case "JSON": self = .json
-    default: return nil
-    }
-  }
-
-  init?(protoName: String) {
```
Comment (NachoSoto) [inline Reference/conformance/conformance.pb.swift]: @allevato how come this is gone now? I was relying on this method myself. Any chance we can bring it back? Thanks!

Comment (allevato) [inline Reference/conformance/conformance.pb.swift]: Can you file an issue describing your use case in detail?

Before the 1.0 milestone we're trying to lock down as much API as we can; it's much easier to add things later than it is to remove or change things once users depend on them. Just as we don't expose the field names as strings publicly, we want to discourage people depending on conversions between the raw enum names and the actual Swift values from outside the standard text format and JSON encoding/decoding processes.

Comment (bordunosp) [reply Reference/conformance/conformance.pb.swift:100]: condemn. I need string value of enum type

**Pull request merged**